### PR TITLE
[exporter/elasticsearch] add data stream routing

### PR DESCRIPTION
**Pull request by andrzej-stencel**
**Description:**

Adds support for routing telemetry based on data stream attributes `data_stream.dataset` and `data_stream.namespace` as described in the Elastic docs: [Data stream naming scheme](https://www.elastic.co/guide/en/fleet/current/data-streams.html#data-streams-naming-scheme).

This PR is based on [[exporter/elasticsearch] Add initial support for metrics #33513](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33513).

**Link to tracking Issue:**

**Testing:**

Added tests.

**Documentation:**

Added documentation for the new configuration options.

### CODE REVIEW

**Comment by felixbarny** [comment]
Let's make sure to also read the data_stream.* properties from resource attributes and scope attributes. The order of precedence is:

1. Signal-level attributes (highest precedence)
2. Scope attributes
3. Resource attributes

**Comment by felixbarny** [comment]
What happens when dynamic indexing in data_stream mode is enabled and the data doesn't contain any data_stream attributes?

What's supposed to happens is that we still add `data_stream.*` attributes to the resulting document.

```
data_stream:
  type: <logs|metrics|traces>
  dataset: generic
  namespace: default  
```

Similar to the [`reroute`](https://www.elastic.co/guide/en/elasticsearch/reference/current/reroute-processor.html) processor, we should also sanitize the values, i.e. replace invalid characters with underscores, and limit the length of dataset and namespace to 100 characters.

See also the code of the `RerouteProcessors` for which characters are disallowed: https://github.com/elastic/elasticsearch/blob/95c9aad670cb8d3429444c197ddfbb0001df025a/modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/RerouteProcessor.java#L205-L206.

File: exporter/elasticsearchexporter/model.go [hunk 1]
```diff
@@ -161,123 +157,31 @@ func (m *encodeModel) encodeLogECSMode(resource pcommon.Resource, record plog.Lo
...
-		} else if m.dedot {
-			doc.Sort()
-		}
-
-		var buf bytes.Buffer
-		err := doc.Serialize(&buf, m.dedot)
-		if err != nil {
-			return nil, err
-		}
+func (m *encodeModel) encodeMetricDataPoint(resource pcommon.Resource, _ pcommon.InstrumentationScope, metric pmetric.Metric, dataPoint pmetric.NumberDataPoint) ([]byte, error) {
```
Comment (felixbarny) [inline exporter/elasticsearchexporter/model.go:168]: From reading the code, I didn't quite understand how grouping metrics with the same attributes to the same ES document works now that the hashing stuff is removed. Does each document contain a single metric/data point now or where is the grouping happening?

From a storage perspective, it's still best to group as many metrics into the same document as possible.

Comment (andrzej-stencel) [reply exporter/elasticsearchexporter/model.go:168]: > Does each document contain a single metric/data point now 

Yes, this is the state after removing hashing. To bring back grouping metrics in a document I would need to reimplement hashing in a different place in code (in exporter and not in  model). I can do this if needed, in this PR or in a separate one.

**Comment by andrzej-stencel** [comment]
Closing this, will open a PR against upstream Contrib repo.

EDIT: Created https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/33755.

File: exporter/elasticsearchexporter/model.go (hunk 1 above)
Comment (felixbarny) [reply exporter/elasticsearchexporter/model.go:168]: Gotcha, let's do this in a follow-up PR then. Can you create an issue so that we can track this?

Comment (andrzej-stencel) [reply exporter/elasticsearchexporter/model.go:168]: Created https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/33756.

**Pull request closed without merging**